### DISCUSSION TITLE
How to dismiss layer on form post response that is redirect?

### DISCUSSION THREAD

**Discussion by jpic** [unpoly/unpoly#400] [Q&A] [answered]
Hi!

Great v2 :tophat: 

I'm integrating with Django framework, and trying to have a form in a modal that is successful to close the modal.

When Django server has valid forms, it typically returns a 302 Redirect response:

https://github.com/django/django/blob/main/django/views/generic/edit.py#L63-L65

However, I'm not finding a way to dismiss the layer in that case... well, no simple/classy way.

Tried adding the X-Up-Dismiss-Layer header to the redirecting response that holds Location, but that didn't do the trick...

Any idea please?

**Comment by adam12** [comment, marked as answer]
You could target the root layer from the `<form>` element, with `[up-layer="root"]`. This attribute is only used for a successful response, IIRC.

Another option, if you know the final URL format, is to indicate it with an `[up-accept-location]` attribute on the `<a>` element that opens the new layer initially. 

I'm presuming the 301 is to a URL that returns a 200 or 201.

**Reply by jpic** [reply]
Thanks! `[up-layer="root"]` is the best solution for me because it reuses the HTML response after the redirect to update the root layer, as such it provides with the best outcome:

- no need to reload with up-on-accept="up.reload()"
- list page in the background is updated
- flash messages work

Love unpoly, keep up the great job! :tophat:

**Comment by triskweline** [comment]
The demo app also has a [companies CRUD](https://demo.unpoly.com/companies) that shows @adam12's technique with `[up-accept-location]`.  Take a closer look at the *New company* button and the green help dot next to it.

**Reply by alnuaimi94** [reply]
up-accept-location doesn't work with django 
and when I used up-layout="root" in form it work but when the form return error validate 
the modal close and return to edit or add page

**Reply by alexandremjacques** [reply]
`up-accept-location` will work if you set it to the final destination, I mean, the page that the user is redirected for. 

Eg. New user form submit -> save user -> redirect to user details. `up-accept-location` should have the user details URL pattern.

**Reply by jpic** [reply]
Thank you!

I'm having an issue with up-accept-location that I can also see on the demo site, or at least, because I'm using a feature that's not in the demo site: flash messages, and, in the current situation, are **impossible to display to the user**:

- user is on a list view, opens a link with `up-layout="new" up-accept-location="/list" href="/create?_next=/list"`
- user submits a valid form, server adds a flash message for this user, and replies with a redirect to the accepted location (passed with the _next argument, implemented server side: the /list view)
- the browser fetches the response for /list, the modal closes, but the list view is not updated

To work around that, we use `up-on-accept="up.reload()"`. But that means the page will be updated from the HTML returned by the second request from up.reload(), and at that point: the flash messages have already been consumed.

Any solution for that please?

**Reply by apollo13** [reply]
Assuming you are willing to use https://unpoly.readthedocs.io/en/latest/index.html flash messages become as simple as:

```python
class UnpolyMessageMiddleware:
    def __init__(self, get_response):
        self.get_response = get_response

    def __call__(self, request):
        response = self.get_response(request)

        messages = get_messages(request)
        # Not sure if messages.used is the proper check, but it does work well enough for my usecase.
        if request.up and not messages.used:
            for message in messages:
                request.up.emit(
                    "django:message",
                    {
                        "safe": isinstance(message.message, SafeData),
                        "tag": message.level_tag,
                        "message": message.message,
                    },
                )

        return response
```
and the following JS:
```javascript
up.on('django:message', (event) => {
    // show the flash message, I am using a modified version of https://github.com/apvarun/toastify-js but you can do whatever you like here
    toast({
        text: event.message,
        escapeMarkup: !event.safe,
        duration: 3000,
        close: true,
        level: event.tag,
        gravity: "top",
        position: "right",
        stopOnFocus: true,
    })
})
```

I'd be open to add something like this to my library if we where to figure out the best way for this.

**Reply by apollo13** [reply]
Given my code with toastify a flash message will look like this:
![Screenshot from 2022-08-01 10-12-00](https://user-images.githubusercontent.com/81547/182103791-d918a62b-2c21-49f6-9790-6f669d56abf5.png)

The main reason for not using the usual Django style (like you can see in the admin) is that the root layer might be hidden. Note that my code might actually not work well with the admin or other normal sites, I simply only tested for my usecases.

**Reply by alnuaimi94** [reply]
> Given my code with toastify a flash message will look like this:
> 
> ![Screenshot from 2022-08-01 10-12-00](https://user-images.githubusercontent.com/81547/182103791-d918a62b-2c21-49f6-9790-6f669d56abf5.png)
> 
> 
> 
> The main reason for not using the usual Django style (like you can see in the admin) is that the root layer might be hidden. Note that my code might actually not work well with the admin or other normal sites, I simply only tested for my usecases.


Hi dear, I'm tried to do your code but nothing happened. can you share full code.

**Reply by jpic** [reply]
Thanks, this would fix the flash message problem, but on form success, the list view still needs to be loaded twice? Once because the browser follows the redirect, and then with `up.reload()` ?

**Reply by apollo13** [reply]
@jpic You are correct, my bad. I saw you accepted `[up-layer="root"]` as solution, but wouldn't that also render on the root layer if the form has errors?

I have another thing that came to mind; the general process in Django goes like this:

 * show a list.
 * click create item.
 * get a new page with a form.
 * on successful submission you get redirect back to the list with a flash message.

With `up-layer="new" up-accept-location="/list"` and my code from above we achieve the following:

 * the item opens in a modal which is nice
 * flash messages work which is great
 * the modal is closed 

Now the one thing missing is updating the root layer which can be achieved via `up.reload()` which requires a second request. Now I am wondering if it would make sense to use something like `up-accept-result-from-response` (better naming required for sure) which would cause the *parent* layer to be updated according to the response.

So taking the example from the unpoly demo:
```html
<a up-layer="new" 
      up-accept-location="/projects/$id"
      up-on-accepted="up.reload('.table', { focus: ':main' })" 
      href="/projects/new">New project</a>
```
one would be able to write:
```html
<a up-layer="new" 
      up-accept-location="/projects/$id"
      up-accept-result-from-response="true"
      up-target=".table"
      href="/projects/new">New project</a>
```
and after unpoly accepted the modal it would update the parent `.table` with data from the response. In essence the updating of the root layer would behave as if one had simply written:
```
<a  up-target=".table"
      href="/list">New project</a>
```

What do you think, would that be a useful addition? Maybe @triskweline can chime in on the technical viability of it all? Maybe this is even achievable nowadays by doing something in JS?

**Reply by apollo13** [reply]
Mhm if `up-on-accepted` where to provide the response (assuming there is one), one could write:
```html
<a up-layer="new" 
      up-accept-location="/projects/$id"
      up-on-accepted="up.render('.table', { document: response.content })" 
      href="/projects/new">New project</a>
```

**Reply by alexandremjacques** [reply]
> @jpic You are correct, my bad. I saw you accepted `[up-layer="root"]` as solution, but wouldn't that also render on the root layer if the form has errors?
> 
> I have another thing that came to mind; the general process in Django goes like this:
> 
>     * show a list.
> 
>     * click create item.
> 
>     * get a new page with a form.
> 
>     * on successful submission you get redirect back to the list with a flash message.
> 
> 
> With `up-layer="new" up-accept-location="/list"` and my code from above we achieve the following:
> 
>     * the item opens in a modal which is nice
> 
>     * flash messages work which is great
> 
>     * the modal is closed
> 
> 
> Now the one thing missing is updating the root layer which can be achieved via `up.reload()` which requires a second request. Now I am wondering if it would make sense to use something like `up-accept-result-from-response` (better naming required for sure) which would cause the _parent_ layer to be updated according to the response.
> 
> So taking the example from the unpoly demo:
> 
> ```
> <a up-layer="new" 
>       up-accept-location="/projects/$id"
>       up-on-accepted="up.reload('.table', { focus: ':main' })" 
>       href="/projects/new">New project</a>
> ```
> 
> one would be able to write:
> 
> ```
> <a up-layer="new" 
>       up-accept-location="/projects/$id"
>       up-accept-result-from-response="true"
>       up-target=".table"
>       href="/projects/new">New project</a>
> ```
> 
> and after unpoly accepted the modal it would update the parent `.table` with data from the response. In essence the updating of the root layer would behave as if one had simply written:
> 
> ```
> <a  up-target=".table"
>       href="/list">New project</a>
> ```
> 
> What do you think, would that be a useful addition? Maybe @triskweline can chime in on the technical viability of it all? Maybe this is even achievable nowadays by doing something in JS?

@triskweline any chance this happens? Would be great.

**Reply by xinha-sh** [reply]
@triskweline Sorry for tagging you again. 

The accepted answer doesn't solve the issue mentioned in the comments above.
Basically the flash message is lost because of up.reload(). 

This is one of the most common use cases:
1. Show a list with create link
2. Link opens a modal with form.
3. When there's an error the modal should remain opened, else it should close and the list should get updated.
4. Flash message should work when list is getting updated.

What's the recommended way to achieve the above UX ?

**Reply by triskweline** [reply]
Three ideas:

### 1. Have the overlay form render into parent layer

Don't use `[up-accept-location]` or `[up-on-accepted]`. Instead the overlay's form uses `<form up-layer="root">` or `<form up-layer="parent root">` (meaning target a parent layer if we're in the overlay, otherwise render to the root layer.

Both makes the response (including the flash) render in the parent layer. Failed submissions will still render in the overlay, since that is controlled by a separate `[up-fail-layer]` attribute.

### 2. Set the flash with JavaScript

Since you already know the acceptance means a newly created project you could have your callback function set the flash:

```html
<a up-layer="new" 
  up-accept-location="..."
  up-on-accepted="reloadWithFlash('.table', 'Project created')">
```

Then implement a callback like this:

```js
async function reloadWithFlash(selector, flash) {
  await up.reload(selector)
  up.element.affix(document.body, '.flash', { text: flash })
}
```

### 3. Transport the flash with an event

After successful submission, have the server emit an `project:created` event with properties like `{ id: 234, flash: 'Project created' }`. The server can emit events using the `X-Up-Events` response header.

Instead of `[up-accept-location]` use `[up-accept-event]`. All event properties become available to the callback:

```html
<a up-layer="new" 
  up-accept-event="project:created"
  up-on-accepted="reloadWithFlash('.table', event.flash)">
```

**Reply by triskweline** [reply]
@apollo13 @alexandremjacques

> one would be able to write:
> 
> ```
> <a up-layer="new" 
>       up-accept-location="/projects/$id"
>       up-accept-result-from-response="true"
>       up-target=".table"
>       href="/projects/new">New project</a>
> ```
>
> and after unpoly accepted the modal it would update the parent .table with data from the response. 

I think a more general solution would be to make the accepting response available to the `up:layer:accepted` event. This way you could render from the HTML without making another request:

```html
<a up-layer="new"
  up-accept-location="..."
  up-on-accepted="up.render('.table', document: event.response.text)">
```

**Reply by alexandremjacques** [reply]
That would be nice.

**Reply by apollo13** [reply]
Yeah, I came to the same conclusion in https://github.com/unpoly/unpoly/discussions/400#discussioncomment-3385711 -- adding the response to the event would be great.

Option number 1 (with `up-fail-layer`) also works nicely but I always seem to forget that the fail layer can be completely different :) What I somewhat dislike about option 1 is that it is not the calling code who decides what goes where but that the form needs some "knowledge" about whether it might be in a modal or not. But I guess in most cases this will be completely fine.

As for the flash-messages themself I think the best approach is to render them via an event & a compiler: https://github.com/unpoly/unpoly/discussions/400#discussioncomment-3295094

**Reply by alexandremjacques** [reply]
Hi!

Is this solution on the roadmap for Unpoly 3? I mean the `document` part in `up-on-accepted="up.render('.table', document: event.response.text)`.

**Reply by adam12** [reply]
I haven't tested it personally, but docs show [`up.render({document: ... })` support](https://unpoly.com/up.render#options.document).

**Reply by alexandremjacques** [reply]
Great. Must have missed.

Thanks.

**Reply by triskweline** [reply]
Watch out:

- `up.render()` does support a `{ document }` option.
- My [earlier suggestion](https://github.com/unpoly/unpoly/discussions/400#discussioncomment-5011005) was to give the `up:layer:accepted` event a `{ response }` property. That is not done yet.

**Reply by alexandremjacques** [reply]
Oh... I see. So I keep my question: any ETA on that? I'm not in a hurry but it would be great in some edge cases here.

**Reply by apollo13** [reply]
While it is true that `document: ` support is there, I don't think the event already has a response attached to it.

EDIT:// Sorry my github didn't update for the new comments.

**Reply by triskweline** [reply]
>  any ETA on that? 

I'm interested in this feature and have worked on it a bit in [this branch](https://github.com/unpoly/unpoly/tree/hk/make-response-available-to-layer-close-events-2). Unfortunately I cannot give ETAs for my open-source work.

**Reply by apollo13** [reply]
Thank you @triskweline for merging the branch :hugs: https://github.com/unpoly/unpoly/commit/a091b2778f520bb2246599fe30ba918900cff44c now is in master and provides the response.

**Reply by triskweline** [reply]
This is now released as [version 3.2.0](https://github.com/unpoly/unpoly/discussions/498).

**Reply by alexandremjacques** [reply]
> > any ETA on that?
> 
> I'm interested in this feature and have worked on it a bit in [this branch](https://github.com/unpoly/unpoly/tree/hk/make-response-available-to-layer-close-events-2). Unfortunately I cannot give ETAs for my open-source work.

Sure. I think I made myself unclear. By ETA I was trying to mean if this was somewhere in the version roadmap. No specific time. My bad.

**Reply by alexandremjacques** [reply]
> This is now released as [version 3.2.0](https://github.com/unpoly/unpoly/discussions/498).

Sweet!! Thanks!!!
